### PR TITLE
[RFC] Added database support for the binary-uuid type

### PR DESCRIPTION
**Pull request by AntonNguyen**
I haven't worked with binary uuids before, so it'd be great if I could get some extra eyes on this. I was just following the thread in #10432, and decided to take a stab at it. 😄 

refs #10432

### CODE REVIEW

File: src/Database/Type/BinaryUuidType.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.0.0
```
Comment (markstory) [inline src/Database/Type/BinaryUuidType.php]: Should be 3.6.0

File: tests/TestCase/Database/Type/BinaryUuidType.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+     */
+    public function testToDatabase()
+    {
+        $value = 'some data';
+        $result = $this->type->toDatabase($value, $this->driver);
+        $this->assertEquals($value, $result);
+
+        $fh = fopen(__FILE__, 'r');
+        $result = $this->type->toDatabase($fh, $this->driver);
+        $this->assertSame($fh, $result);
```
Comment (markstory) [inline tests/TestCase/Database/Type/BinaryUuidType.php]: Should there be a test for a string uuid coming back as bytes?

**Review by markstory — COMMENTED** [review]
Thanks for taking this on. A few people should be quite happy🎉

**Comment by markstory** [comment]
Updated with schema reflection for each db driver and integrated the new type into the Type mappings. In MySQL, SQLServer and SQLite `BINARY(16)` will now map to `binaryuuid`. Postgres does not support binary columns with lengths so binaryuuid maps to the `UUID` datatype there.

**Comment by markstory** [comment]
Thanks @AntonNguyen :clap:

**Comment by dereuromark** [comment]
@AntonNguyen Do you agree on the following simpifications and fixes? https://github.com/dereuromark/cakephp-shim/pull/11 We could make a follow up PR for 3.6 then.

**Comment by dereuromark** [comment]
Also, when adding invalid UUIDs like `'test'` it will save them as `0e000000000000000000000000000000
` instead of failing, and also a lot of

    Warning (2): pack(): Type H: illegal hex digit t [ROOT/.../BinaryUuidType.php, ...]

etc
But maybe this should be part of the validation/patching part and is out of scope of the type class then?

**Comment by markstory** [comment]
> But maybe this should be part of the validation/patching part and is out of scope of the type class then?

Yes that seems like a validation concern to me.

**Comment by lorenzo** [comment]
I just saw this... would have preferred a sorted binary uuid implementation, like the one we have in some of the tests.

I guess that could be another custom type :)

**Comment by dereuromark** [comment]
Do you mean https://github.com/alizain/ulid or https://github.com/mbbender/doctrine-ordered-uuid-generator/blob/master/src/ORM/Id/OrderedGuidGenerator.php @lorenzo  ?

**Comment by markstory** [comment]
@lorenzo We can still change this implementation as it is in the unreleased 3.next branch.

**Comment by dereuromark** [comment]
Is there any interest to finish this for 3.x or 4.x?

**Pull request merged**